clouddriver-cf/src/main/groovy/com/netflix/spinnaker/clouddriver/cf/provider/agent/ClusterCachingAgent.groovy

```diff
@@ -0,0 +1,504 @@
+import com.netflix.spinnaker.clouddriver.cf.CloudFoundryCloudProvider
+import com.netflix.spinnaker.clouddriver.cf.cache.Keys
+import com.netflix.spinnaker.clouddriver.cf.config.CloudFoundryConstants
+import com.netflix.spinnaker.clouddriver.cf.model.CloudFoundryService
+import com.netflix.spinnaker.clouddriver.cf.provider.CloudFoundryProvider
+import com.netflix.spinnaker.clouddriver.cf.provider.ProviderUtils
+import com.netflix.spinnaker.clouddriver.cf.security.CloudFoundryAccountCredentials
+import com.netflix.spinnaker.clouddriver.cf.utils.CloudFoundryClientFactory
+import org.cloudfoundry.client.lib.CloudFoundryException
+import org.cloudfoundry.client.lib.CloudFoundryOperations
+import org.cloudfoundry.client.lib.domain.*
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+
+import static com.netflix.spinnaker.cats.agent.AgentDataType.Authority.AUTHORITATIVE
+import static com.netflix.spinnaker.clouddriver.cf.cache.Keys.Namespace.*
+
+/**
+ * @author Greg Turnquist
+ */
+class ClusterCachingAgent implements CachingAgent, OnDemandAgent, AccountAware {
+
+  final static Logger log = LoggerFactory.getLogger(ClusterCachingAgent)
+
+  @Deprecated
+  private static final String LEGACY_ON_DEMAND_TYPE = 'CloudFoundry' // not relevant for CF
+
+  private static final String ON_DEMAND_TYPE = 'ServerGroup'
+
+  private static final TypeReference<Map<String, Object>> ATTRIBUTES = new TypeReference<Map<String, Object>>() {}
+
+  static final Set<AgentDataType> types = Collections.unmodifiableSet([
+      AUTHORITATIVE.forType(SERVER_GROUPS.ns),
+      AUTHORITATIVE.forType(CLUSTERS.ns),
+      AUTHORITATIVE.forType(APPLICATIONS.ns),
+      AUTHORITATIVE.forType(INSTANCES.ns),
+      AUTHORITATIVE.forType(LOAD_BALANCERS.ns)
+  ] as Set)
+
+  final CloudFoundryCloudProvider cloudFoundryCloudProvider
+  final CloudFoundryClientFactory cloudFoundryClientFactory
+  final CloudFoundryAccountCredentials account
+  final ObjectMapper objectMapper
+  final Registry registry
+
+  final OnDemandMetricsSupport metricsSupport
+
+  ClusterCachingAgent(CloudFoundryCloudProvider cloudFoundryCloudProvider,
+                      CloudFoundryClientFactory cloudFoundryClientFactory,
+                      CloudFoundryAccountCredentials account,
+                      ObjectMapper objectMapper,
+                      Registry registry) {
+    this.objectMapper = objectMapper
+    this.account = account
+    this.cloudFoundryClientFactory = cloudFoundryClientFactory
+    this.cloudFoundryCloudProvider = cloudFoundryCloudProvider
+    this.registry = registry
+    this.metricsSupport = new OnDemandMetricsSupport(registry, this, cloudFoundryCloudProvider.id + ":" + ON_DEMAND_TYPE)
+  }
+
+  @Override
+  String getProviderName() {
+    CloudFoundryProvider.PROVIDER_NAME
+  }
+
+  @Override
+  String getAgentType() {
+    "${account.name}/${account.org}/${account.space}/${ClusterCachingAgent.simpleName}"
```

Comment:
I don't think it hurts anything to do so, but you don't need to partition your agentType on org and space if there is only 1 agent per account (in AWS there is one agent per account+region)

in the CF credentials model can there be multiple org/space associated with one account?